[LogisticRegression] Match Spark CPU behaviors when dataset has one label

lijinf2 opened the pull request.

lijinf2 commented:
build

eordentlich commented on python/src/spark_rapids_ml/classification.py:
```diff
@@ -976,6 +976,12 @@ def _single_fit(init_parameters: Dict[str, Any]) -> Dict[str, Any]:
+                if len(logistic_regression.classes_) == 1:
+                    if init_parameters["fit_intercept"] is True:
+                        model["coef_"] = [[0.0] * logistic_regression.n_cols]
+                        model["intercept_"] = [float("inf")]
```
Would the sign of this depend on the label value?

lijinf2 replied:
Revised to support -inf for label 0.

eordentlich commented on python/src/spark_rapids_ml/classification.py:
```diff
@@ -1027,6 +1033,16 @@ def _out_schema(self) -> Union[StructType, str]:
+        if len(result["classes_"]) == 1:
+            if self.getFitIntercept() is False:
+                print(
+                    "WARNING: All labels belong to a single class and fitIntercept=false. It's a dangerous ground, so the algorithm may not converge."
```
Can we match spark's warning?

lijinf2 replied:
Is there a way to capture spark scala warning in python? I tried caplog.set_level() to INFO, WARN, CRITICAL but got empty log text.

lijinf2 replied:
Revised to use logger.warning

eordentlich commented on python/src/spark_rapids_ml/classification.py:
```diff
@@ -1027,6 +1033,16 @@ def _out_schema(self) -> Union[StructType, str]:
+                )
+            else:
+                print(
+                    "WARNING: All labels are the same value and fitIntercept=true, so the coefficients will be zeros. Training is not needed."
```
Same here.

lijinf2 replied:
Revised

eordentlich commented on python/tests/test_logistic_regression.py:
```diff
@@ -1076,3 +1076,71 @@ def test_parameters_validation() -> None:
+    )
+    num_rows = len(X)
+
+    weight = np.ones([num_rows])
```
any reason to have weight column since it is not supported?

lijinf2 replied:
removed

eordentlich commented on python/tests/test_logistic_regression.py:
```diff
@@ -1076,3 +1076,71 @@ def test_parameters_validation() -> None:
+            assert blor_model.intercept == 0.0
+        else:
+            assert array_equal(blor_model.coefficients.toArray(), [0, 0], 0.0)
+            assert blor_model.intercept == float("inf")
```
Maybe check what happens also in the case if all labels are 0 instead of 1 (i.e. y).

lijinf2 replied:
added

eordentlich commented on python/src/spark_rapids_ml/classification.py:
```diff
@@ -976,6 +976,18 @@ def _single_fit(init_parameters: Dict[str, Any]) -> Dict[str, Any]:
+                    class_val = logistic_regression.classes_[0]
+                    assert (
+                        class_val == 1.0 or class_val == 0.0
+                    ), "class value must be either 1. or 0. when dataset has one label"
```
What does spark do if label has one value but is not 1 or 0?

lijinf2 replied:
Revised. 
if label < 0, a java runtimeError pops up. 
If label > 1, spark trains a multinomial classification, cuml trains a single-class classification due to using y.unique().

lijinf2 commented:
build

eordentlich commented on python/tests/test_logistic_regression.py:
```diff
@@ -1076,3 +1076,102 @@ def test_parameters_validation() -> None:
+        blor_model = blor.fit(bdf)
+
+        if fit_intercept is False:
+            if label == 1.0:
```
Can caplog be used to check warning in this case?  Like here: https://github.com/NVIDIA/spark-rapids-ml/blob/branch-23.12/python/tests/test_nearest_neighbors.py#L47

eordentlich commented:
Any update on this?

eordentlich commented:
You will probably need to patch the ci docker image in this pr to get tests to pass as rapidsai-nightly no longer has cuml 23.12.   switch to rapidsai channel.

lijinf2 commented:
build

lijinf2 commented:
> You will probably need to patch the ci docker image in this pr to get tests to pass as rapidsai-nightly no longer has cuml 23.12. switch to rapidsai channel.

Added the caplog, and a test case to check invalid label. Just updated ci docker image and yes seems ci can run.

lijinf2 commented:
build

eordentlich reviewed (approved):
👍
